tests/integration/ha/test_scaling.py

```diff
@@ -0,0 +1,217 @@
+        - [ 'sysctl', '-w', 'vm.swappiness=0' ]
+        - [ 'sysctl', '-w', 'net.ipv4.tcp_retries2=5' ]
+    """,
+}
+TLS_CERTIFICATES_APP_NAME = "self-signed-certificates"
+
+HTTP_UNITS = [0, 1, 2]
+HTTPS_UNITS = [3, 4, 5]
+
+APP_AND_TLS = [APP_NAME, TLS_CERTIFICATES_APP_NAME]
+
+
+@pytest.mark.group(1)
+@pytest.mark.skip_if_deployed
+@pytest.mark.abort_on_fail
+@pytest.mark.charm
+async def test_build_and_deploy(ops_test: OpsTest):
+    """Deploying all charms required for the tests, and wait for their complete setup to be done."""
+
+    charm = await ops_test.build_charm(".")
+    await ops_test.model.deploy(charm, application_name=APP_NAME, num_units=1)
+
+    # Opensearch
+    await ops_test.model.set_config(OPENSEARCH_CONFIG)
+    # NOTE: can't access 2/stable from the tests, only 'edge' available
+    await ops_test.model.deploy(OPENSEARCH_APP_NAME, channel="2/edge", num_units=1)
+
+    config = {"ca-common-name": "CN_CA"}
+    await ops_test.model.deploy(TLS_CERTIFICATES_APP_NAME, channel="stable", config=config)
+
+    await ops_test.model.wait_for_idle(
+        apps=[TLS_CERTIFICATES_APP_NAME], status="active", timeout=1000
+    )
+
+    # Relate it to OpenSearch to set up TLS.
+    await ops_test.model.integrate(OPENSEARCH_APP_NAME, TLS_CERTIFICATES_APP_NAME)
+    await ops_test.model.wait_for_idle(
+        apps=[OPENSEARCH_APP_NAME, TLS_CERTIFICATES_APP_NAME], status="active", timeout=1000
+    )
+
+    async with ops_test.fast_forward():
+        await ops_test.model.wait_for_idle(
+            apps=[APP_NAME], wait_for_exact_units=1, timeout=1000, idle_period=30
+        )
+
+    assert ops_test.model.applications[APP_NAME].status == "blocked"
+
+    pytest.relation = await ops_test.model.integrate(OPENSEARCH_APP_NAME, APP_NAME)
+    await ops_test.model.wait_for_idle(
+        apps=[OPENSEARCH_APP_NAME, APP_NAME], status="active", timeout=1000
+    )
+
+
+##############################################################################
+# Helper functions
+##############################################################################
+
+
+async def scale_up(ops_test: OpsTest, amount: int, https: bool = False) -> None:
+    """Testing that newly added units are functional."""
+    init_units_count = len(ops_test.model.applications[APP_NAME].units)
+    expected = init_units_count + amount
+
+    # scale up
+    logger.info(f"Adding {amount} units")
+    await ops_test.model.applications[APP_NAME].add_unit(count=amount)
+
+    logger.info(f"Waiting for {amount} units to be added and stable")
+    await ops_test.model.wait_for_idle(
+        apps=[APP_NAME],
+        status="active",
+        wait_for_exact_units=expected,
+        timeout=1000,
+        idle_period=30,
+    )
+
+    num_units = len(ops_test.model.applications[APP_NAME].units)
+    assert num_units == expected
+
+    logger.info("Checking the functionality of the new units")
+    assert await access_all_dashboards(ops_test, pytest.relation.id, https)
```

Comment:
Would be very helpful to rename this function to `access_all_dashboards_units` - so it is clear that this helper checks all units